File: 2 Add Two Numbers/2add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        def func1(l):
            if l.next == None:
                return l.val
            return l.val + func1(l.next) * 10 

        def func2(val):
            if val:
                return ListNode(val = val % 10, next = func2(val // 10))
            return None
        
        l1 = func2(func1(l1) + func1(l2))
        return l1 if l1 else ListNode(val=0)
            

File: 2 Add Two Numbers/2add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int s = 0;
        ListNode* res = new ListNode();
        ListNode* dummy = res;

        while (l1 != nullptr || l2 != nullptr or s != 0){
            
            if (l1 != nullptr){
                s += l1 -> val;
                l1 = l1 -> next;
            }

            if (l2 != nullptr){
                s += l2 -> val;
                l2 = l2 -> next;
            }

            res -> next = new ListNode(s % 10);
            s /= 10;
            res = res -> next;
        }
        return dummy -> next;
    }
};